src/algebra/star/self_adjoint.lean

```diff
@@ -0,0 +1,118 @@
+/-
+Copyright (c) 2021 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+
+import algebra.star.basic
+import group_theory.subgroup.basic
+
+/-!
+# Self-adjoint elements of a star additive group
+
+This file defines `self_adjoint R`, where `R` is a star additive monoid, as the additive subgroup
+containing the elements that satisfy `star x = x`. This includes, for instance, Hermitian
+operators on Hilbert spaces.
+
+## TODO
+
+* If `R` is a `star_module R₂ R`, put a module structure on `self_adjoint R`. This would naturally
+be a `module (self_adjoint R₂) (self_adjoint R)`, but doing this literally would be undesirable
+since in the main case of interest (`R₂ = ℂ`) we want `module ℝ (self_adjoint R)` and not
+`module (self_adjoint ℂ) (self_adjoint R)`. One way of doing this would be to add the typeclass
+`[has_trivial_star R]`, of which `ℝ` would be an instance, and then add a
+`[module R (self_adjoint E)]` instance whenever we have `[module R E] [has_trivial_star E]`. Another
+one would be to define a `[star_invariant_scalars R E]` to express the fact that
+`star (x • v) = x • star v`.
+
+-/
+
+variables {R : Type*}
+
+variables (R)
+/-- The self-adjoint elements of a star additive group, as an additive subgroup. -/
+def self_adjoint [add_group R] [star_add_monoid R] : add_subgroup R :=
+{ carrier := {x | star x = x},
+  zero_mem' := star_zero R,
+  add_mem' := λ x y (hx : star x = x) (hy : star y = y),
+                show star (x + y) = x + y, by simp only [star_add x y, hx, hy],
+  neg_mem' := λ x (hx : star x = x), show star (-x) = -x, by simp only [hx, star_neg] }
+variables {R}
+
+namespace self_adjoint
+
+section add_group
+variables [add_group R] [star_add_monoid R]
+
+lemma mem_iff {x : R} : x ∈ self_adjoint R ↔ star x = x :=
+by { rw [←add_subgroup.mem_carrier], exact iff.rfl }
+
+instance : has_star (self_adjoint R) := ⟨id⟩
+instance : has_involutive_star (self_adjoint R) := ⟨λ _, rfl⟩
+
+@[simp] lemma star_eq {x : self_adjoint R} : star x = x := rfl
+@[simp] lemma star_coe_eq {x : self_adjoint R} : star (x : R) = x := x.prop
+
+instance : star_add_monoid (self_adjoint R) := ⟨λ x y, star_eq⟩
+
+end add_group
+
+instance [add_comm_group R] [star_add_monoid R] : add_comm_group (self_adjoint R) :=
+{ add_comm := add_comm,
+  ..add_subgroup.to_add_group (self_adjoint R) }
+
+section ring
+variables [ring R] [star_ring R]
+
+instance : has_one (self_adjoint R) := ⟨⟨1, by rw [mem_iff, star_one]⟩⟩
+
+@[simp] lemma coe_one : (coe : self_adjoint R → R) (1 : self_adjoint R) = (1 : R) := rfl
+
+lemma one_mem : (1 : R) ∈ self_adjoint R := by simp only [mem_iff, star_one]
+
+lemma conjugate {x : R} (hx : x ∈ self_adjoint R) {z : R} : z * x * star z ∈ self_adjoint R :=
```

Comment:
Could you please add a TODO mentioning that it would be nice someday to make this more abstract, by defining `z * x * star z` as a monoid action of `R` on `R` (similar to the existing `conj_act`), and then stating this lemma as the fact that `self_adjoint` is invariant under it.

Reply:
Also, I think this could usefully have `z` explicit.

Reply:
Done!